Fixes #28541 - Fix host creation from image and hostgroup

shiramax opened the pull request:
In the code we are translating the forman image id to the provider image_id and put it under ['host']['compute_attributes']['image_id'], 
the problem is that we relied on compute_resource_id to exist or host_id (in case of update) and we didn't address a commend that send only with hostgroup, 
this PR should fix the issue.

kgaikwad commented:
@ofedoren, any updates on this?

rahulbajaj0509 commented on lib/hammer_cli_foreman/hosts/common_update_options.rb:
```diff
@@ -100,7 +100,13 @@ def request_params
+            compute_resource_id = ::HammerCLIForeman::ComputeResources.get_host_compute_resource_id(params['id'])
+          elsif params['host']['hostgroup_id']
+            compute_resource_id = ::HammerCLIForeman::ComputeResources.get_hostgroup_compute_resource_id(params['host']['hostgroup_id'])
+          end
```
Nit: Sorry for adding this nit, it is a personal choice and feel free to ignore :)
When we have multiple if/else/ifelse, we can use the `case` statement. Something like:

```
if options['option_image_id']
  compute_resource_id = case [params['host']['compute_resource_id'], ::HammerCLIForeman::ComputeResources.get_host_compute_resource_id(params['id'])]
                          when params['host']['compute_resource_id'] then params['host']['compute_resource_id']
                          when params['id'] then ::HammerCLIForeman::ComputeResources.get_host_compute_resource_id(params['id'])
                          when params['host']['hostgroup_id'] then ::HammerCLIForeman::ComputeResources.get_hostgroup_compute_resource_id(params['host']['hostgroup_id'])
                        end
  raise ArgumentError, "Missing argument for 'compute_resource'" if compute_resource_id.nil?
  image_uuid =  ::HammerCLIForeman::ComputeResources.get_image_uuid(compute_resource_id, options["option_image_id"])
  params['host']['compute_attributes']['image_id'] = image_uuid
end

I haven't tested this, but it looks like it. @shiramax @kgaikwad let me know your thoughts on this!

shiramax replied:
thanks @rahulbajaj0509 , Done :)

kgaikwad replied:
@rahulbajaj0509, @shiramax 

In my opinion, +1 to if/else/ifelse loop that was there previously.
It was more readable.

kgaikwad commented on lib/hammer_cli_foreman/hosts/common_update_options.rb:
```diff
@@ -98,9 +98,12 @@ def request_params
-        if options["option_image_id"]
-          compute_resource_id = params['host']['compute_resource_id'] || ::HammerCLIForeman::ComputeResources.get_host_compute_resource_id(params['id'])
+        if options['option_image_id']
+          compute_resource_id = case
```
@shiramax, 

There is no variable or regex inside case statement then against which value `when` statements will do matching? 
You need to mention something inside case statement otherwise it will take first value from first when statement.
Please correct me if I am wrong.

kgaikwad replied:
For readability, it would be good to have simple `if/else/ifelse` as I mentioned.

shiramax replied:
@kgaikwad 
from what I understand it possible to create a switch case statement without the case like follows: 

case
when a < 3
  "Smaller than 3"
end

and I also tested it, and it works fine to me, can you confirm it works for you as well ?

ofedoren commented on lib/hammer_cli_foreman/hosts/common_update_options.rb:
```diff
@@ -98,9 +98,14 @@ def request_params
+        if options['option_image_id']
+          if params['host']['compute_resource_id']
+            compute_resource_id = params['host']['compute_resource_id']
+          elsif  params['id']
```
Nit: extra space after `elsif`.

ofedoren commented on lib/hammer_cli_foreman/compute_resource/utils.rb:
```diff
@@ -8,6 +8,14 @@ def self.get_image_uuid(compute_resource_id, image_id)
+      HammerCLIForeman.record_to_common_format(
+          HammerCLIForeman.foreman_resource(:hostgroups).call(
+              :show, 'id' => hostgroup_id
+          )['compute_resource_id']
```
After some testing I've noticed, that if we use host group that is nested then the server response contains only the name of a compute resource (`"compute_resource_id" => nil`), so this will cause error `Missing argument for 'compute_resource'`.

What do you think will be better:
  1. Support resource id resolving in case we have the resource name only?
  2. Fix this on API side.

shiramax replied:
@ofedoren good catch! thanks @ofedoren.
I'm working on a fix for that in foreman-core. I will update this one as soon as I finish with the PR .
 thanks

shiramax commented:
@ofedoren thanks !
I fixed the issue in hammer, because it's faster but I've opened another for foreman, and when he will ve merged I will change it here as well. 
I've also added some test :) 
It would be great if you can review it today so we would be able to release a new hammer version today. 
thanks :)

tbrisker commented on lib/hammer_cli_foreman/compute_resource/utils.rb:
```diff
@@ -8,6 +8,21 @@ def self.get_image_uuid(compute_resource_id, image_id)
+          HammerCLIForeman.foreman_resource(:hostgroups).call(:show, 'id' => hostgroup_id)
+      )
+      compute_resource_id = hostgroup['compute_resource_id']
+      unless compute_resource_id
```
should this be something like?
```suggestion
      if hostgroup['compute_resource_name'].present? && compute_resource_id.nil?
```
so we don't do the extra query if there is no compute resource?

kgaikwad replied:
In case no ActiveSupport extension, then
```suggestion
      if hostgroup['compute_resource_name'].to_s.strip.empty? && compute_resource_id.nil?
```

tbrisker commented on lib/hammer_cli_foreman/compute_resource/utils.rb:
```diff
@@ -8,6 +8,21 @@ def self.get_image_uuid(compute_resource_id, image_id)
+      unless compute_resource_id
+
+        compute_resources = HammerCLIForeman.record_to_common_format(
+            HammerCLIForeman.foreman_resource(:compute_resources).call(:index)
```
should we add a search for the name instead of getting the entire list (which might be paginated) and looking for the id in it?

shiramax commented:
@tbrisker @kgaikwad done, thanks!

tbrisker commented on lib/hammer_cli_foreman/compute_resource/utils.rb:
```diff
@@ -8,6 +8,21 @@ def self.get_image_uuid(compute_resource_id, image_id)
+            HammerCLIForeman.foreman_resource(:compute_resources).call(
+                :index, :search => "name = \"#{hostgroup['compute_resource_name']}\""
+            )
+         )['results'][0]['id']
```
what if no matching resource are found?  `['results'][0]` will be `nil` i think?

shiramax replied:
@tbrisker fixed :)
